src/main/java/software/amazon/ionpathextraction/PathExtractorImpl.java

```diff
@@ -80,27 +76,28 @@ public void match(final IonReader reader, final T context) {
 
     private int matchRecursive(final IonReader reader, final Tracker<T> tracker, final T context) {
         final int currentDepth = tracker.getCurrentDepth();
-        int currentPosition = 0;
+        int readerContainerIndex = 0;
 
         while (reader.next() != null) {
             // will continue to next depth
             final List<SearchPath<T>> partialMatches = new ArrayList<>();
+            final String[] annotations = reader.getTypeAnnotations();
 
             for (SearchPath<T> sp : tracker.activePaths()) {
-                boolean match = searchPathPartialMatchAt(sp, reader, tracker.getCurrentDepth(), currentPosition);
-                boolean isTerminal = isTerminal(tracker.getCurrentDepth(), sp);
-
-                if (match && isTerminal) {
-                    int stepOutTimes = invokeCallback(reader, sp, tracker.getInitialReaderDepth(), context);
-                    if (stepOutTimes > 0) {
-                        return stepOutTimes - 1;
+                final MatchContext matchContext = new MatchContext(reader, currentDepth, readerContainerIndex, config);
+                boolean spTerminal = sp.isTerminal(tracker.getCurrentDepth());
+                boolean partialMatch = sp.partialMatchAt(matchContext);
+
+                if (partialMatch) {
+                    if (spTerminal) {
```

Comment:
* What does `spTerminal` indicate?
* I think you should expand `sp` to `searchPath` for readability.

Reply:
will do. Will add some javadoc to SearchPath that will make more clear what being terminal means